x-pack/plugin/data-streams/src/main/java/org/elasticsearch/xpack/datastreams/UpdateTimeSeriesRangeService.java

```diff
@@ -0,0 +1,170 @@
+                onComplete.run();
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                LOGGER.warn("failed to update tsdb data stream end times", e);
+                onComplete.run();
+            }
+
+        }, ClusterStateTaskExecutor.unbatched());
+    }
+
+    void setPollInterval(TimeValue newValue) {
+        LOGGER.info(
+            "updating ["
+                + DataStreamsPlugin.TIME_SERIES_POLL_INTERVAL.getKey()
+                + "] setting from ["
+                + pollInterval
+                + "] to ["
+                + newValue
+                + "]"
+        );
+        this.pollInterval = newValue;
+    }
+
+    ClusterState updateTimeSeriesTemporalRange(ClusterState current, Instant now) {
+        Metadata.Builder mBuilder = null;
+        for (DataStream dataStream : current.metadata().dataStreams().values()) {
+            if (dataStream.getIndexMode() != IndexMode.TIME_SERIES) {
+                continue;
+            }
+            if (dataStream.isReplicated()) {
+                continue;
+            }
+
+            // getWriteIndex() selects the latest added index:
+            String head = dataStream.getWriteIndex().getName();
+            IndexMetadata im = current.metadata().getIndices().get(head);
+            Instant currentEnd = IndexSettings.TIME_SERIES_END_TIME.get(im.getSettings());
+            TimeValue lookAheadTime = IndexSettings.LOOK_AHEAD_TIME.get(im.getSettings());
+            Instant newEnd = now.plus(lookAheadTime.getMillis(), ChronoUnit.MILLIS).plus(pollInterval.getMillis(), ChronoUnit.MILLIS);
+            if (newEnd.isAfter(currentEnd)) {
+                Settings settings = Settings.builder()
+                    .put(IndexSettings.TIME_SERIES_END_TIME.getKey(), DEFAULT_DATE_TIME_FORMATTER.format(newEnd))
+                    .build();
+                LOGGER.debug(
+                    "updating [{}] setting from [{}] to [{}] for index [{}]",
+                    IndexSettings.TIME_SERIES_END_TIME.getKey(),
+                    currentEnd,
+                    newEnd,
+                    head
+                );
+                if (mBuilder == null) {
+                    mBuilder = Metadata.builder(current.metadata());
+                }
+                mBuilder.updateSettings(settings, head);
+                Metadata.Builder finalMBuilder = mBuilder;
+                // Verify that all temporal ranges of each backing index is still valid:
+                dataStream.validate(finalMBuilder::get);
+            }
+        }
+
+        if (mBuilder != null) {
+            return ClusterState.builder(current).metadata(mBuilder).build();
+        } else {
+            return current;
+        }
+    }
+
+    void scheduleTask() {
+        if (job == null) {
+            LOGGER.debug("schedule tsdb update task");
+            job = threadPool.schedule(() -> perform(this::scheduleTask), pollInterval, ThreadPool.Names.SAME);
```

Comment:
We schedule the thread here on the `SAME` threadpool, which *think* becomes the `MANAGEMENT` thread (correct me if I am wrong), but in the even that it is rejected in the future (for example, if we were to bound the queue size of the management thread pool, or if we were to use a different threadpool entirely) then the `onComplete.run()` in `clusterStateProcessed` and `onFailure` inside the `perform` method would throw an exception.

We should avoid throwing exceptions there, so I think we should wrap the invocations of `onComplete.run()` in a try/catch.

The downside of doing the reschedule this way (which you already explained your reasoning for) is that if for any reason the scheduling fails, it won't be rescheduled until a new master is elected. We should try to be paranoid about any case where it could accidentally get "unscheduled" (I can't see any right now other than the `.schedule` call throwing an exception, but maybe I'm just paranoid about it).

Reply:
Also, should we switch to `scheduleUnlessShuttingDown` instead of `schedule` here?

Reply:
I will change this to use `scheduleWithFixedDelay(...)`, so that we avoid failing to manually re-schedule in case of a failure to invoke `schedule(...)` or a bug. In order to fix my own paranoia with not scheduling to to update tsdb data streams multiple times, I will add a running flag to keep track whether we are performing the updates to tsdb data streams.